Batteries/Control/LawfulMonadState.lean

```diff
@@ -0,0 +1,225 @@
+/-
+Copyright (c) 2025 Devon Tuma. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Devon Tuma, Quang Dao
+-/
+
+/-!
+# Laws for Monads with State
+
+This file defines a typeclass for `MonadStateOf` with compatible `get` and `set` operations.
+
+Note that we use `MonadStateOf` over `MonadState` as the first induces the second,
+but we phrase things using `MonadStateOf.set` and `MonadState.get` as those are the
+versions that are available at the top level namespace.
+-/
+
+/-- The namespaced `MonadStateOf.get` is equal to the `MonadState` provided `get`. -/
+@[simp] theorem monadStateOf_get_eq_get [MonadStateOf σ m] :
+    (MonadStateOf.get : m σ) = get := rfl
+
+/-- The namespaced `MonadStateOf.modifyGet` is equal to the `MonadState` provided `modifyGet`. -/
+@[simp] theorem monadStateOf_modifyGet_eq_modifyGet [MonadStateOf σ m]
+    (f : σ → α × σ) : (MonadStateOf.modifyGet f : m α) = modifyGet f := rfl
+
+/-- Class for well behaved state monads, extending the base `MonadState` type.
+Requires that `modifyGet` is equal to the same definition with only `get` and `set`,
+that `get` is idempotent if the result isn't used, and that `get` after `set` returns
+exactly the value that was previously `set`. -/
+class LawfulMonadStateOf (σ : Type _) (m : Type _ → Type _) [Monad m] [MonadStateOf σ m] where
+  /-- `modifyGet f` is equal to getting the state, modifying it, and returning a result. -/
+  modifyGet_eq {α} (f : σ → α × σ) :
+    modifyGet (m := m) f = do let z ← f <$> get; set z.2; return z.1
+  /-- Discarding the result of `get` is the same as never getting the state. -/
+  get_bind_const {α} (mx : m α) : (do let _ ← get; mx) = mx
+  /-- Calling `get` twice is the same as just using the first retreived state value. -/
+  get_bind_get_bind {α} (mx : σ → σ → m α) :
+    (do let s ← get; let s' ← get; mx s s') = (do let s ← get; mx s s)
+  /-- Setting the monad state to its current value has no effect. -/
+  get_bind_set_bind {α} (mx : σ → PUnit → m α) :
+    (do let s ← get; let u ← set s; mx s u) = (do let s ← get; mx s PUnit.unit)
+  /-- Setting and then returning the monad state is the same as returning the set value. -/
+  set_bind_get (s : σ) : (do set (m := m) s; get) = (do set s; return s)
+  /-- Setting the monad twice is the same as just setting to the final state. -/
+  set_bind_set (s s' : σ) : (do set (m := m) s; set s') = set s'
+
+variable {σ : Type _} {m : Type _ → Type _} [Monad m] [MonadStateOf σ m] [LawfulMonadStateOf σ m]
```

Comment:
`variable` declarations outside a section or namespace are best avoided in Batteries. Can you localize these to one of the namespaces below? (If there are universe issues, they should be explained and documented.)